src/bootstrap/src/core/build_steps/compile.rs

```diff
@@ -190,10 +192,19 @@ impl Step for Std {
 
         let mut target_deps = builder.ensure(StartupObjects { compiler, target });
 
-        let compiler_to_use = builder.compiler_for(compiler.stage, compiler.host, target);
+        let mut compiler_to_use = builder.compiler_for(compiler.stage, compiler.host, target);
         trace!(?compiler_to_use);
 
-        if compiler_to_use != compiler {
+        if compiler_to_use != compiler
+            // Never uplift std unless we have compiled stage 2; if stage 2 is compiled,
+            // uplift it from there.
+            //
+            // FIXME: improve `fn compiler_for` to avoid adding stage condition here.
+            && compiler.stage > 2
+        {
+            if compiler.stage > 2 {
```

Comment:
nit: this is already checked 2 lines above

Reply:
Oops, I must forgot updating this part while resolving merge conflicts.